Added warning for unused arithmetic expressions

aaronaaeng opened the pull request:
The compiler now displays a warning when a binary arithmetic operation is evaluated but not used.  This resolves #50124  by following the instructions outlined in the issue.  The changes are as follows:

- Added new pattern matching for unused arithmetic expressions in `src/librustc_lint/unused.rs`
- Added `#[must_use]` attributes to the binary operation methods in `src/libcore/internal_macros.rs`
- Added `#[must_use]` attributes to the non-assigning binary operators in `src/libcore/ops/arith.rs`

scottmcm commented on src/libcore/internal_macros.rs:
```diff
@@ -38,6 +38,7 @@ macro_rules! forward_ref_binop {
     };
     (impl $imp:ident, $method:ident for $t:ty, $u:ty, #[$attr:meta]) => {
         #[$attr]
+        #[must_use]
```
Is the `impl` the right place for this attribute?  I thought the lint only looked at methods on trait definitions.

varkor replied:
Sorry, I didn't read the mentoring instructions before, but this fix needs tweaking:
- Instead of adding `#[must_use]` to method definitions in impls, you need to add them to the trait definitions (e.g. here: https://github.com/rust-lang/rust/blob/699eb95ae3dae023cddca85facf6660b196bd0d1/src/libcore/ops/arith.rs#L97).
- You don't need any of the `#[must_use]` in the macros, for the same reason.

aaronaaeng replied:
I don't fully understand what is going on in the `impl`, though I'm a little unsure of where else I would put the attribute.  The recurrence of `#[$attr]` in each `impl` made it seem as though the attributes need to be repeated before each possible combination of `T` and `U` states.

aaronaaeng replied:
Oh, ok.  I can go fix that.  Thanks for the quick support.  I also just learned that even though GitHub will automatically show new messages as they come in, I need to refresh to see edits.

varkor replied:
Take a look at `PartialOrd`, which implements `lt` (`<`): https://github.com/rust-lang/rust/blob/699eb95ae3dae023cddca85facf6660b196bd0d1/src/libcore/cmp.rs#L642-L650
The `#[must_use]` attribute is on the definition in `trait PartialOrd`, but not on any of the `impl PartialOrd`. The way `#[must_use]` works at the moment, it's only the trait declaration which is important — everything else inherits from the trait. (That way, _anyone_ who implements an operator will get the advantage of `#[must_use]` without having to add it themselves.

scottmcm commented on src/librustc_lint/unused.rs:
```diff
@@ -103,6 +103,11 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
                         op_warned = true;
                     },
+                    hir::BiAdd | hir::BiSub | hir::BiDiv | hir::BiMul | hir::BiRem => {
+                        let msg = "unused arithmetic operation which must be used";
```
1. While you're here, maybe add the bitops too?  `a^b;` giving a warning also makes sense to me
2. Unary ops too? `-x` seems to also fit "unused arithmetic operation which must be used"
3. At that point, is it just all ops?  The only three left are `*`, `&&`, and `||`.  (The short-circuiting ones I could see an argument for not having the lint , since `foo() || panic!()` is possible, but...)

aaronaaeng replied:
I'm working on the `UnaryOp` pattern matching.  While matching on binary operators, it matches on `bin_op.node`.  However, the `UnOp` type doesn't have this field.  I've tried to find the definition of this type but I've come up with nothing so far.  Any suggestion as to what file I should check out?

abonander replied:
`UnOp` is an enum [defined here](https://github.com/rust-lang/rust/blob/master/src/librustc/hir/mod.rs#L1054). `BinOp` is an enum as well but it's [wrapped in a struct that gives it span information](https://github.com/rust-lang/rust/blob/master/src/librustc/hir/mod.rs#L1054), which I assume was considered unnecessary for unary ops.

All the `hir::` types are defined in `src/librust/hir.mod.rs`. For convenience all the enums have their variants reexported in this module as well.

varkor commented on src/librustc_lint/unused.rs:
```diff
@@ -103,6 +103,11 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
                         cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
                         op_warned = true;
                     },
+                    hir::BiAdd | hir::BiSub | hir::BiDiv | hir::BiMul | hir::BiRem => {
```
Yeah, different messages sound good 👍
We want the remaining operators too, though (the same message as the other non-comparison operators is fine).

aaronaaeng commented:
I made the changes discussed.  Warnings are now thrown whenever an operator is evaluated but not used.  The warning is specific to the type of operator that went unused.  The `#[must_use]` attribute has been moved to the correct location for each operator.  It passes all of the `x.py` tests and works as intended with sample code that evaluates expressions using binary and unary operators but does not use the results.

abonander commented:
Putting `#[must_use]` on the traits themselves is probably the right move but I don't know what the policy is for expanding the scope of lints. If someone has `#[deny(unused_must_use)]` in their crate and they made a mistake with one of these operators, this change will break their build. It's a *good* breakage, but a breakage nonetheless.

abonander commented on src/librustc_lint/unused.rs:
```diff
@@ -103,11 +103,33 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
+                    },
+                }
+            }
+            if let hir::ExprUnary(un_op, ..) = expr.node {
```
I would combine this and the `if let hir::ExprBinary` bit above into a single `match`.

aaronaaeng replied:
What is the best practice for combining these two match statements?  Combining them seems like it would require nested matches since `ExprBinary` and `ExprUnary` are different types.  I have to match on `op.node` for binary operators but only on `op` for unary operators.  Should I implement this nested system or is there a better way?

abonander replied:
A nested match is fine. It's mostly about eliminating the redundant code here.

abonander commented on src/librustc_lint/unused.rs:
```diff
@@ -103,11 +103,33 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
+                    hir::BiAdd | hir::BiSub | hir::BiDiv | hir::BiMul | hir::BiRem => {
+                        let msg = "unused arithmetic operation which must be used";
+                        cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
+                        op_warned = true;
```
I would extract the above two lines out of each arm and just have each arm return its relevant message and then emit the lint at the end.

abonander commented on src/librustc_lint/unused.rs:
```diff
@@ -103,11 +103,33 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
+                    hir::BiAnd | hir::BiOr => {
+                        let msg = "unused logical operation which must be used";
+                        cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
+                        op_warned = true;
```
For bonus points you could emit a help message here saying to use `if (first_op) { second_op }` instead, but that's a bit involved. I might tack that on myself if you don't want to get into it.

abonander commented:
I don't see a lot of UI tests for lints in general so I don't know how we want to test these additions.

zackmdavis commented:
> Putting `#[must_use]` on the traits themselves is probably the right move but

Note that `#[must_use]` for functions is still unstable. (Although probably not for long: the stabilization final comment period [ended](https://github.com/rust-lang/rust/issues/43302#issuecomment-380311079) with no objections, and who can say but that I might even have time to rebase the [stabilization PR](https://github.com/rust-lang/rust/pull/48925) tomorrow??)

> I don't know what the policy is for expanding the scope of lints

Given that [RFC 1193](https://github.com/rust-lang/rfcs/blob/master/text/1193-cap-lints.md) lint capping exists specifically to preven the breakage from spreading to dependent crates, it's probably OK sometimes (but is definitely OK here given `fn_must_use` not even being stable). (See #48717 for an example of Crater-detected breakage from a lint behavior change.)

> I don't see a lot of UI tests for lints in general

There's a src/test/ui/lint directory, but this would probably go with rust/src/test/ui/rfc_1940-must_use_on_functions/fn_must_use.rs (which should be moved to src/test/ui/lint/ on stabilization).

abonander commented:
@zackmdavis This change also affects the arithmetic, the logical/bitwise operators and the unary operators. Where would those lint tests go.

zackmdavis commented:
new file `src/test/ui/lint/must_use_ops.rs`??

aaronaaeng commented:
I updated the match statements to reduce redundant code.  Because I am not incredibly familiar with pattern matching, I introduced a bool that says whether to send the lint message.  Without this check, it could throw empty warnings on expressions that were correct.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/369670631) ([raw log](https://api.travis-ci.org/v3/job/369670631/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:05:51] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:05:51] tidy error: /checkout/src/librustc_lint/unused.rs:95: trailing whitespace
[00:05:51] tidy error: /checkout/src/librustc_lint/unused.rs:98: trailing whitespace
[00:05:51] tidy error: /checkout/src/librustc_lint/unused.rs:99: trailing whitespace
[00:05:51] tidy error: /checkout/src/librustc_lint/unused.rs:115: trailing whitespace
[00:05:53] some tidy checks failed
[00:05:53] 
[00:05:53] 
[00:05:53] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:05:53] 
[00:05:53] 
[00:05:53] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:05:53] Build completed unsuccessfully in 0:02:33
[00:05:53] Build completed unsuccessfully in 0:02:33
[00:05:53] Makefile:79: recipe for target 'tidy' failed
[00:05:53] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:136cc378
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

zackmdavis commented on src/librustc_lint/unused.rs:
```diff
@@ -91,23 +93,39 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
-                    _ => {},
+            match expr.node {
+                hir::ExprBinary(bin_op, ..)  => {
+                    should_warn_op = true;
```
```rust
let must_use_op = match expr.node {
    hir::ExprBinary(bin_op, ..)  => {
        match bin_op.node {
            hir::BiEq | hir::BiLt | hir::BiLe | hir::BiNe | hir::BiGe | hir::BiGt => {
                Some("comparison")
            },
            hir::BiAdd | hir::BiSub | hir::BiDiv | hir::BiMul | hir::BiRem => {
                Some("arithmetic operation")
            },
            hir::BiAnd | hir::BiOr => {
                Some("logical operation")
            },
            hir::BiBitXor | hir::BiBitAnd | hir::BiBitOr | hir::BiShl | hir::BiShr => {
                Some("bitwise operation")
            },
        }
    },
    hir::ExprUnary(..) => {
        Some("unary operation")
    },
    _ => None
}
if let Some(must_use_op) {
    cx.span_lint(UNUSED_MUST_USE, expr.span, format!("unused {} which must be used", must_use_op));
    op_warned = true;
}
```

aaronaaeng commented:
After seeing that, it seemed pretty obvious.  I made the necessary changes and it is a lot cleaner than it was.  Hopefully it all looks good now.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/369763475) ([raw log](https://api.travis-ci.org/v3/job/369763475/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:04:56] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:04:57] tidy error: /checkout/src/librustc_lint/unused.rs:117: line longer than 100 chars
[00:04:58] some tidy checks failed
[00:04:58] 
[00:04:58] 
[00:04:58] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:58] 
[00:04:58] 
[00:04:58] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:58] Build completed unsuccessfully in 0:01:59
[00:04:58] Build completed unsuccessfully in 0:01:59
[00:04:58] make: *** [tidy] Error 1
[00:04:58] Makefile:79: recipe for target 'tidy' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:1254f026
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

zackmdavis commented on src/librustc_lint/unused.rs:
```diff
@@ -91,23 +91,34 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
+                        },
+                    }
+                },
+                hir::ExprUnary(..) => {
```
I just realized we can save two lines by writing this as `hir::ExprUnary(..) => Some("unary operation"),` rather than with a block (obviously this is a criticism of my earlier review comment, not you)

zackmdavis commented:
Needs [tests](https://rust-lang-nursery.github.io/rustc-guide/tests/adding.html) (probably a new UI test file at src/test/ui/lint/must_use_ops.rs or similar; don't forget the `#[feature(fn_must_use)]`), 

For the benefit of future software-archeologists, I argue that it's also worth [squashing](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) these 6+ commits down into 1. (GitHub truncates the commit message summary line at 72 characters, so something like "make must-use lint cover binary operations" (42 chars) is better than 12d227aff's 95 chars.)

aaronaaeng commented:
Should I include each operator or just one from each category, i.e., do I need to include an expected warning for `+`, `-`, `*`, and `/`, or should I just include one operator from the arithmetic category?

aaronaaeng commented:
My attempt at a rebase did not fix the message as hoped.  Let me try again.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/369881072) ([raw log](https://api.travis-ci.org/v3/job/369881072/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:05:42] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:05:43] tidy error: /checkout/src/test/ui/lint/must-use-ops.rs: missing trailing newline
[00:05:43] tidy error: /checkout/src/librustc_lint/unused.rs:96: trailing whitespace
[00:05:44] some tidy checks failed
[00:05:44] 
[00:05:44] 
[00:05:44] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:05:44] 
[00:05:44] 
[00:05:44] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:05:44] Build completed unsuccessfully in 0:02:32
[00:05:44] Build completed unsuccessfully in 0:02:32
[00:05:44] make: *** [tidy] Error 1
[00:05:44] Makefile:79: recipe for target 'tidy' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0021e6f5
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

zackmdavis commented:
looks good to me (but I don't have the authority to approve PRs)

zackmdavis commented:
>  wants to merge 1 commit into rust-lang:master from aaronaaeng:master

(Standard practice is to start a new branch in your own repository, rather than working on master; this surprised me just now as I began to rebase #48925 off this.)

aaronaaeng commented:
That makes sense.  I'll do that next time I submit a PR somewhere.  Thank you for helping me out with all of this.

estebank commented:
@bors r+

bors commented:
:pushpin: Commit 91aa267 has been approved by `estebank`

<!-- @bors r=estebank 91aa267eda21be36d7cae43b6809eba63f1aac4c -->

kennytm commented:
@bors p=6

bors commented:
:hourglass: Testing commit 91aa267eda21be36d7cae43b6809eba63f1aac4c with merge 1eb0cef62b3e83bf7ce6bc7d1b99e130648a55e8...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.7186), [status-travis](https://travis-ci.org/rust-lang/rust/builds/372310179?utm_source=github_status&utm_medium=notification)
Approved by: estebank
Pushing 1eb0cef62b3e83bf7ce6bc7d1b99e130648a55e8 to master...

leonardo-m commented:
In the last Nightly I still don't see those warnings. Perhaps I am missing something?

kennytm commented:
@leonardo-m You need `#![feature(fn_must_use)]` to see those warnings for now. After #48925 is merged (within 1 to 2 days) they should be enabled automatically.
